src/modules/rlm_magic/rlm_magic.c

```diff
@@ -0,0 +1,249 @@
+	CONF_PARSER_TERMINATOR};
+
+/*
+ *	Do any per-module initialization that is separate to each
+ *	configured instance of the module.  e.g. set up connections
+ *	to external databases, read configuration files, set up
+ *	dictionary entries, etc.
+ */
+static int mod_instantiate(UNUSED CONF_SECTION *conf, void *instance)
+{
+	rlm_magic_t *inst = instance;
+	DEBUG2("Creating Magic module");
+
+	if (inst->is_local)
+	{
+		#ifndef HAVE_SYS_UN_H
+			ERROR("Trying to use local unix sockets on a system that does not support it");
+			return -1;
+		#endif
+		DEBUG2("Using local Unix socket %s for authentication", inst->socket_address);
+	}
+	else
+	{
+		DEBUG2("Using remote server %s:%i for authentication", inst->socket_address, inst->port);
+	}
+
+	return 0;
+}
+
+/*
+ *	Don't currently do anything here
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authorize(UNUSED void *instance, UNUSED REQUEST *request)
+{
+	return RLM_MODULE_NOOP;
+}
+
+/*
+ *	Authenticate the user with the given password.
+ */
+static rlm_rcode_t CC_HINT(nonnull) mod_authenticate(void *instance, REQUEST *request)
+{
+	// The instance variables
+	rlm_magic_t *inst = instance;
+
+	//Socket related vars
+	char buffer[BUFSIZ];
+	int sockfd;
+
+	DEBUG2("Attempting Magic Authentication");
+
+	if (inst->is_local)
+	{
+		sockfd = fr_socket_client_unix(inst->socket_address, false);
+		if (sockfd == -1)
+		{
+			fr_perror("socket");
+			return RLM_MODULE_FAIL;
+		}
+	}
+	else
+	{
+		fr_ipaddr_t address_info;
+		address_info.af = AF_INET;
+		struct hostent *hostent;
+		
+		hostent = gethostbyname(inst->socket_address);
+		if (hostent == NULL)
+		{
+			fprintf(stderr, "error: gethostbyname(\"%s\")\n", inst->socket_address);
+			return RLM_MODULE_FAIL;
+		}
+		in_addr_t in_addr = inet_addr(inet_ntoa(*(struct in_addr *)*(hostent->h_addr_list)));
+		if (in_addr == (in_addr_t)-1)
+		{
+			fprintf(stderr, "error: inet_addr(\"%s\")\n", *(hostent->h_addr_list));
+			return RLM_MODULE_FAIL;
+		}
+		address_info.ipaddr.ip4addr.s_addr = in_addr;
+
+		sockfd = fr_socket_client_tcp(NULL, &address_info, inst->port, false);
+	}
+	// Create our message which needs to be in this format
+	// '{"address": "username", "password": "password", "sessionId": ""}\n'
+	char *socket_message = talloc_array(request, char, BUFSIZ);
+	strlcpy(socket_message, "{\"address\": \"", BUFSIZ);
+	strlcat(socket_message, request->username->vp_strvalue, BUFSIZ);
+	strlcat(socket_message, "\", \"password\": \"", BUFSIZ);
+	strlcat(socket_message, request->password->vp_strvalue, BUFSIZ);
+	strlcat(socket_message, "\", \"sessionId\": \"\"}\n", BUFSIZ);
+
+	// Write to the socket
+	if (write(sockfd, socket_message, BUFSIZ) == -1)
+	{
+		perror("write");
+		return RLM_MODULE_FAIL;
+	}
+	if (shutdown(sockfd, SHUT_WR) != 0)
+	{
+		perror("shutdown socket");
+		return RLM_MODULE_FAIL;
+	}
+	talloc_free(socket_message);
+
+	// read from the socket
+	int nbytes_read = read(sockfd, buffer, BUFSIZ);
```

Comment:
If you're only reading 1 byte from the buffer, there's no need to read `BUFSIZE` bytes.  That will generally block the server on the `read()` call.  The only way to stop that is if the other end either writes `BUFSIZE` data, OR writes less than `BUFSIZE` bytes, and then closes the socket.